File: src/devices/simulatedbike/ridescenario.cpp
#include "ridescenario.h"

#include <cstdio>
#include <cstdlib>
#include <fstream>
#include <sstream>

namespace {

const RideValue absentValue = {false, 0.0};

std::string trim(const std::string &s) {
    const std::string ws = " \t\r\n";
    const std::string::size_type first = s.find_first_not_of(ws);
    if (first == std::string::npos)
        return std::string();
    const std::string::size_type last = s.find_last_not_of(ws);
    return s.substr(first, last - first + 1);
}

/** Everything from an unquoted # to the end of the line is a comment. */
std::string stripComment(const std::string &line) {
    const std::string::size_type hash = line.find('#');
    if (hash == std::string::npos)
        return line;
    return line.substr(0, hash);
}

/**
 * Parse a number, rejecting anything with trailing rubbish. strtod alone accepts "90rpm" as
 * 90, which turns a typo into a plausible ride.
 */
bool parseNumber(const std::string &s, double *out) {
    if (s.empty())
        return false;
    const char *begin = s.c_str();
    char *end = nullptr;
    const double v = strtod(begin, &end);
    if (end == begin || *end != '\0')
        return false;
    *out = v;
    return true;
}

/**
 * Bounds exist to catch typos, not to model physiology: a missing decimal point or a
 * transposed pair of digits should stop the file rather than produce a ride nobody notices is
 * wrong. They are deliberately far wider than any real session.
 */
struct FieldBound {
    const char *name;
    double min;
    double max;
};

bool checkBound(const FieldBound &b, double v, std::string *error, int lineNo) {
    if (v >= b.min && v <= b.max)
        return true;
    if (error) {
        std::ostringstream os;
        os << "line " << lineNo << ": " << b.name << " of " << v << " is outside " << b.min
           << ".." << b.max;
        *error = os.str();
    }
    return false;
}

void formatNumber(std::ostringstream &os, double v) {
    // Canonical form drops a trailing ".0" so that a hand-written "watts=200" survives a
    // round trip looking like itself. Anything fractional keeps enough digits to be exact
    // for the resolutions the format actually carries.
    if (v == static_cast<long long>(v)) {
        os << static_cast<long long>(v);
        return;
    }
    char buf[32];
    snprintf(buf, sizeof(buf), "%g", v);
    os << buf;
}

void appendField(std::ostringstream &os, const char *name, const RideValue &f) {
    if (!f.present)
        return;
    os << ' ' << name << '=';
    formatNumber(os, f.value);
}

/**
 * One field's timeline, evaluated at t: interpolate between the two nearest samples that both
 * state it, hold outside them, absent before the first.
 *
 * Linear search over the samples rather than an index. A scenario is tens of lines and this is
 * called at the player's tick rate; an index would be a structure to keep in step for no
 * measurable gain.
 */
RideValue valueAt(const std::vector<RideSample> &samples, RideValue RideSample::*field, double t) {
    const RideSample *before = nullptr;
    const RideSample *after = nullptr;

    for (std::vector<RideSample>::const_iterator it = samples.begin(); it != samples.end(); ++it) {
        if (!((*it).*field).present)
            continue;
        if (it->t <= t)
            before = &(*it);
        else if (!after)
            after = &(*it);
    }

    if (!before) {
        // Before the first statement of this field the bike has said nothing about it. Holding
        // the *first* value backwards would invent a reading the ride never had.
        return absentValue;
    }
    if (!after || after->t == before->t)
        return ((*before).*field);

    const double span = after->t - before->t;
    const double frac = (t - before->t) / span;
    const double a = ((*before).*field).value;
    const double b = ((*after).*field).value;
    const RideValue r = {true, a + (b - a) * frac};
    return r;
}

} // namespace

RideScenario::RideScenario()
    : m_valid(false), m_mode(RIDE_MODE_POWER), m_startResistance(-1),
      m_ergLag(RideScenario::defaultErgLag()), m_noise(RideScenario::defaultNoise()) {}

RideScenario RideScenario::parse(const std::string &text, std::string *error) {
    RideScenario s;
    if (error)
        error->clear();

    static const FieldBound boundWatts = {"watts", 0.0, 3000.0};
    static const FieldBound boundCadence = {"cadence", 0.0, 250.0};
    static const FieldBound boundHr = {"hr", 0.0, 250.0};
    static const FieldBound boundSpeed = {"speed", 0.0, 150.0};
    static const FieldBound boundResistance = {"resistance", 0.0, 254.0};

    std::istringstream in(text);
    std::string rawLine;
    int lineNo = 0;
    bool modeSeen = false;

    while (std::getline(in, rawLine)) {
        ++lineNo;
        const std::string line = trim(stripComment(rawLine));
        if (line.empty())
            continue;

        // A sample line is any line whose first token starts with "t=". Everything else is a
        // directive, which keeps the two kinds of line impossible to confuse.
        if (line.compare(0, 2, "t=") == 0) {
            RideSample sample;
            sample.t = 0;
            sample.watts = absentValue;
            sample.cadence = absentValue;
            sample.hr = absentValue;
            sample.speed = absentValue;
            sample.resistance = absentValue;
            sample.silence = 0;
            bool haveT = false;

            std::istringstream ls(line);
            std::string token;
            while (ls >> token) {
                const std::string::size_type eq = token.find('=');
                if (eq == std::string::npos || eq == 0) {
                    if (error) {
                        std::ostringstream os;
                        os << "line " << lineNo << ": '" << token << "' is not key=value";
                        *error = os.str();
                    }
                    return RideScenario();
                }
                const std::string key = token.substr(0, eq);
                double v = 0;
                if (!parseNumber(token.substr(eq + 1), &v)) {
                    if (error) {
                        std::ostringstream os;
                        os << "line " << lineNo << ": '" << token << "' has no usable number";
                        *error = os.str();
                    }
                    return RideScenario();
                }

                const RideValue set = {true, v};
                if (key == "t") {
                    if (v < 0) {
                        if (error) {
                            std::ostringstream os;
                            os << "line " << lineNo << ": t must not be negative";
                            *error = os.str();
                        }
                        return RideScenario();
                    }
                    sample.t = v;
                    haveT = true;
                } else if (key == "watts") {
                    if (!checkBound(boundWatts, v, error, lineNo))
                        return RideScenario();
                    sample.watts = set;
                } else if (key == "cadence") {
                    if (!checkBound(boundCadence, v, error, lineNo))
                        return RideScenario();
                    sample.cadence = set;
                } else if (key == "hr") {
                    if (!checkBound(boundHr, v, error, lineNo))
                        return RideScenario();
                    sample.hr = set;
                } else if (key == "speed") {
                    if (!checkBound(boundSpeed, v, error, lineNo))
                        return RideScenario();
                    sample.speed = set;
                } else if (key == "resistance") {
                    if (!checkBound(boundResistance, v, error, lineNo))
                        return RideScenario();
                    sample.resistance = set;
                } else if (key == "silence") {
                    if (v <= 0) {
                        if (error) {
                            std::ostringstream os;
                            os << "line " << lineNo << ": silence must be positive";
                            *error = os.str();
                        }
                        return RideScenario();
                    }
                    sample.silence = v;
                } else {
                    if (error) {
                        std::ostringstream os;
                        os << "line " << lineNo << ": unknown field '" << key << "'";
                        *error = os.str();
                    }
                    return RideScenario();
                }
            }

            if (!haveT) {
                if (error) {
                    std::ostringstream os;
                    os << "line " << lineNo << ": sample has no t";
                    *error = os.str();
                }
                return RideScenario();
            }
            if (!s.m_samples.empty() && sample.t <= s.m_samples.back().t) {
                if (error) {
                    std::ostringstream os;
                    os << "line " << lineNo << ": t=" << sample.t << " does not advance past "
                       << s.m_samples.back().t;
                    *error = os.str();
                }
                return RideScenario();
            }
            s.m_samples.push_back(sample);
            continue;
        }

        // Directive: name, then one value.
        std::istringstream ds(line);
        std::string key;
        std::string value;
        ds >> key >> value;
        if (value.empty()) {
            if (error) {
                std::ostringstream os;
                os << "line " << lineNo << ": directive '" << key << "' has no value";
                *error = os.str();
            }
            return RideScenario();
        }
        std::string extra;
        if (ds >> extra) {
            if (error) {
                std::ostringstream os;
                os << "line " << lineNo << ": directive '" << key << "' has more than one value";
                *error = os.str();
            }
            return RideScenario();
        }

        if (key == "mode") {
            if (value == "power")
                s.m_mode = RIDE_MODE_POWER;
            else if (value == "resistance")
                s.m_mode = RIDE_MODE_RESISTANCE;
            else if (value == "speed")
                s.m_mode = RIDE_MODE_SPEED;
            else {
                if (error) {
                    std::ostringstream os;
                    os << "line " << lineNo << ": mode must be power, resistance or speed, not '"
                       << value << "'";
                    *error = os.str();
                }
                return RideScenario();
            }
            modeSeen = true;
        } else if (key == "bike") {
            s.m_bike = value;
        } else if (key == "resistance") {
            double v = 0;
            if (!parseNumber(value, &v) || !checkBound(boundResistance, v, error, lineNo)) {
                if (error && error->empty()) {
                    std::ostringstream os;
                    os << "line " << lineNo << ": resistance '" << value << "' is not a number";
                    *error = os.str();
                }
                return RideScenario();
            }
            s.m_startResistance = v;
        } else if (key == "erg_lag") {
            double v = 0;
            if (!parseNumber(value, &v) || v < 0) {
                if (error) {
                    std::ostringstream os;
                    os << "line " << lineNo << ": erg_lag '" << value
                       << "' is not a number of seconds";
                    *error = os.str();
                }
                return RideScenario();
            }
            s.m_ergLag = v;
        } else if (key == "noise") {
            double v = 0;
            if (!parseNumber(value, &v) || v < 0 || v > 1) {
                if (error) {
                    std::ostringstream os;
                    os << "line " << lineNo << ": noise '" << value
                       << "' is not a fraction between 0 and 1";
                    *error = os.str();
                }
                return RideScenario();
            }
            s.m_noise = v;
        } else {
            if (error) {
                std::ostringstream os;
                os << "line " << lineNo << ": unknown directive '" << key << "'";
                *error = os.str();
            }
            return RideScenario();
        }
    }

    if (!modeSeen) {
        if (error)
            *error = "no mode directive: state power, resistance or speed";
        return RideScenario();
    }
    if (s.m_samples.empty()) {
        if (error)
            *error = "no samples";
        return RideScenario();
    }

    s.m_valid = true;
    return s;
}

RideScenario RideScenario::load(const std::string &path, std::string *error) {
    std::ifstream f(path.c_str());
    if (!f) {
        if (error)
            *error = "cannot open " + path;
        return RideScenario();
    }
    std::ostringstream buf;
    buf << f.rdbuf();
    RideScenario s = parse(buf.str(), error);
    if (!s.valid() && error && !error->empty())
        *error = path + ": " + *error;
    return s;
}

double RideScenario::duration() const {
    if (m_samples.empty())
        return 0;
    const RideSample &last = m_samples.back();
    return last.t + last.silence;
}

RidePoint RideScenario::at(double t) const {
    RidePoint p;
    p.t = t;
    p.watts = absentValue;
    p.cadence = absentValue;
    p.hr = absentValue;
    p.speed = absentValue;
    p.resistance = absentValue;
    p.silent = false;

    if (!m_valid)
        return p;

    // A silence window swallows everything inside it, including any sample that would fall in
    // the middle of one. Reporting a value from within a dropout is the bug the dropout
    // scenario exists to catch, so the check comes first and returns.
    for (std::vector<RideSample>::const_iterator it = m_samples.begin(); it != m_samples.end();
         ++it) {
        if (it->silence > 0 && t >= it->t && t < it->t + it->silence) {
            p.silent = true;
            return p;
        }
    }

    p.watts = valueAt(m_samples, &RideSample::watts, t);
    p.cadence = valueAt(m_samples, &RideSample::cadence, t);
    p.hr = valueAt(m_samples, &RideSample::hr, t);
    p.speed = valueAt(m_samples, &RideSample::speed, t);
    p.resistance = valueAt(m_samples, &RideSample::resistance, t);
    return p;
}

std::string RideScenario::toText() const {
    std::ostringstream os;
    if (!m_valid)
        return std::string();

    os << "mode        ";
    switch (m_mode) {
    case RIDE_MODE_POWER:
        os << "power";
        break;
    case RIDE_MODE_RESISTANCE:
        os << "resistance";
        break;
    case RIDE_MODE_SPEED:
        os << "speed";
        break;
    }
    os << '\n';

    if (!m_bike.empty())
        os << "bike        " << m_bike << '\n';
    if (m_startResistance >= 0) {
        os << "resistance  ";
        formatNumber(os, m_startResistance);
        os << '\n';
    }
    if (m_ergLag != defaultErgLag()) {
        os << "erg_lag     ";
        formatNumber(os, m_ergLag);
        os << '\n';
    }
    if (m_noise != defaultNoise()) {
        os << "noise       ";
        formatNumber(os, m_noise);
        os << '\n';
    }
    os << '\n';

    for (std::vector<RideSample>::const_iterator it = m_samples.begin(); it != m_samples.end();
         ++it) {
        os << "t=";
        formatNumber(os, it->t);
        appendField(os, "watts", it->watts);
        appendField(os, "cadence", it->cadence);
        appendField(os, "hr", it->hr);
        appendField(os, "speed", it->speed);
        appendField(os, "resistance", it->resistance);
        if (it->silence > 0) {
            os << " silence=";
            formatNumber(os, it->silence);
        }
        os << '\n';
    }

    return os.str();
}

File: tools/fakebike-android/app/src/main/java/org/qzfork/fakebike/RideScenario.java
package org.qzfork.fakebike;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.List;

/**
 * A ride scenario, ported from src/devices/simulatedbike/ridescenario.cpp.
 *
 * This is the third player of the same format - the simulated bike inside QZ plays it, the
 * Layer C tests play it, and this peripheral plays it out over a real radio. The same file
 * through all three has to produce the same numbers, so the two rules that are easy to get
 * wrong are ported exactly rather than approximated:
 *
 * <p><b>Each field is its own timeline.</b> A value interpolates between the two nearest
 * samples that both state it, is absent before the first that states it, and holds after the
 * last. So {@code hr} can appear in one sample in ten without dragging its neighbours to
 * zero, and {@code cadence=0} is an explicit value to be reproduced exactly rather than read
 * as "said nothing".
 *
 * <p><b>Interpolation is always linear.</b> No step mode, no easing, no per-field override.
 *
 * <p>{@code silence=<seconds>} is the exception that is not about values: the bike reports
 * nothing at all for that long. On this peripheral that is a genuine gap in the notification
 * stream, which is the one thing the DIRCON loop in the test suite cannot produce.
 *
 * <p>Kept deliberately dependency-free and mechanical. If the C++ changes, this has to change
 * with it, and a reader comparing the two should be able to do it line by line.
 */
public class RideScenario {

    public static final int MODE_POWER = 0;
    public static final int MODE_RESISTANCE = 1;
    public static final int MODE_SPEED = 2;

    /** One field of one sample. {@code present == false} means the line did not state it. */
    public static class Value {
        public final boolean present;
        public final double value;

        Value(boolean present, double value) {
            this.present = present;
            this.value = value;
        }
    }

    private static final Value ABSENT = new Value(false, 0);

    /** One line of the file. */
    private static class Sample {
        double t;
        Value watts = ABSENT;
        Value cadence = ABSENT;
        Value hr = ABSENT;
        Value speed = ABSENT;
        Value resistance = ABSENT;
        double silence;
    }

    /** Every field at one instant, as a player should report it. */
    public static class Point {
        public double t;
        public Value watts = ABSENT;
        public Value cadence = ABSENT;
        public Value hr = ABSENT;
        public Value speed = ABSENT;
        public Value resistance = ABSENT;
        /** True while a silence window is open: the bike reports nothing at all. */
        public boolean silent;
    }

    private boolean valid;
    private String name = "";
    private int mode = MODE_POWER;
    private String bike = "";
    private double startResistance = -1;
    private double ergLag = 2.0;
    private double noise = 0.0;
    private final List<Sample> samples = new ArrayList<>();

    public boolean isValid() { return valid; }
    public String getName() { return name; }
    public int getMode() { return mode; }
    public String getBike() { return bike; }
    public double getStartResistance() { return startResistance; }
    public double getErgLag() { return ergLag; }
    public double getNoise() { return noise; }

    public double duration() {
        double last = 0;
        for (Sample s : samples) {
            if (s.t > last) last = s.t;
            if (s.t + s.silence > last) last = s.t + s.silence;
        }
        return last;
    }

    public static RideScenario load(InputStream in, String name) throws IOException {
        StringBuilder text = new StringBuilder();
        BufferedReader reader =
                new BufferedReader(new InputStreamReader(in, Charset.forName("UTF-8")));
        String line;
        while ((line = reader.readLine()) != null) {
            text.append(line).append('\n');
        }
        reader.close();
        RideScenario s = parse(text.toString());
        s.name = name;
        return s;
    }

    public static RideScenario parse(String text) {
        RideScenario s = new RideScenario();
        for (String raw : text.split("\n")) {
            String line = stripComment(raw).trim();
            if (line.isEmpty()) continue;

            if (line.startsWith("t=") || line.startsWith("t =")) {
                Sample sample = parseSample(line);
                if (sample == null) return s; // invalid: valid stays false
                s.samples.add(sample);
                continue;
            }

            String[] parts = line.split("\\s+", 2);
            if (parts.length < 2) return s;
            String key = parts[0];
            String value = parts[1].trim();
            try {
                if (key.equals("mode")) {
                    if (value.equals("power")) s.mode = MODE_POWER;
                    else if (value.equals("resistance")) s.mode = MODE_RESISTANCE;
                    else if (value.equals("speed")) s.mode = MODE_SPEED;
                    else return s;
                } else if (key.equals("bike")) {
                    s.bike = value;
                } else if (key.equals("resistance")) {
                    s.startResistance = Double.parseDouble(value);
                } else if (key.equals("erg_lag")) {
                    s.ergLag = Double.parseDouble(value);
                } else if (key.equals("noise")) {
                    s.noise = Double.parseDouble(value);
                } else {
                    return s; // an unknown directive is a typo, not something to ignore
                }
            } catch (NumberFormatException e) {
                return s;
            }
        }
        s.valid = !s.samples.isEmpty();
        return s;
    }

    private static String stripComment(String line) {
        int hash = line.indexOf('#');
        return hash >= 0 ? line.substring(0, hash) : line;
    }

    private static Sample parseSample(String line) {
        Sample sample = new Sample();
        boolean haveT = false;
        for (String token : line.split("\\s+")) {
            int eq = token.indexOf('=');
            if (eq <= 0) return null;
            String key = token.substring(0, eq).trim();
            String rest = token.substring(eq + 1).trim();
            double v;
            try {
                v = Double.parseDouble(rest);
            } catch (NumberFormatException e) {
                return null;
            }
            if (key.equals("t")) {
                sample.t = v;
                haveT = true;
            } else if (key.equals("watts")) {
                sample.watts = new Value(true, v);
            } else if (key.equals("cadence")) {
                sample.cadence = new Value(true, v);
            } else if (key.equals("hr")) {
                sample.hr = new Value(true, v);
            } else if (key.equals("speed")) {
                sample.speed = new Value(true, v);
            } else if (key.equals("resistance")) {
                sample.resistance = new Value(true, v);
            } else if (key.equals("silence")) {
                sample.silence = v;
            } else {
                return null;
            }
        }
        return haveT ? sample : null;
    }

    /** Which field of a sample a timeline is being built from. */
    private interface Field {
        Value of(Sample s);
    }

    private static final Field WATTS = new Field() { public Value of(Sample s) { return s.watts; } };
    private static final Field CADENCE = new Field() { public Value of(Sample s) { return s.cadence; } };
    private static final Field HR = new Field() { public Value of(Sample s) { return s.hr; } };
    private static final Field SPEED = new Field() { public Value of(Sample s) { return s.speed; } };
    private static final Field RESISTANCE =
            new Field() { public Value of(Sample s) { return s.resistance; } };

    private Value valueAt(Field field, double t) {
        Sample before = null;
        Sample after = null;
        for (Sample s : samples) {
            if (!field.of(s).present) continue;
            if (s.t <= t) before = s;
            else if (after == null) after = s;
        }
        if (before == null) {
            // Before the first statement of this field the bike has said nothing about it.
            // Holding the first value backwards would invent a reading the ride never had.
            return ABSENT;
        }
        if (after == null || after.t == before.t) return field.of(before);

        double span = after.t - before.t;
        double frac = (t - before.t) / span;
        double a = field.of(before).value;
        double b = field.of(after).value;
        return new Value(true, a + (b - a) * frac);
    }

    public Point at(double t) {
        Point p = new Point();
        p.t = t;
        if (!valid) return p;

        // A silence window swallows everything inside it, including a sample that would fall
        // in the middle of one. Reporting a value from within a dropout is the bug the
        // dropout scenario exists to catch, so this check comes first and returns.
        for (Sample s : samples) {
            if (s.silence > 0 && t >= s.t && t < s.t + s.silence) {
                p.silent = true;
                return p;
            }
        }

        p.watts = valueAt(WATTS, t);
        p.cadence = valueAt(CADENCE, t);
        p.hr = valueAt(HR, t);
        p.speed = valueAt(SPEED, t);
        p.resistance = valueAt(RESISTANCE, t);
        return p;
    }
}
